cuqi/experimental/mcmc/_hmc.py

```diff
@@ -0,0 +1,401 @@
+
+        # initialization
+        j, s, n = 0, 1, 1
+        point_minus, point_plus = np.copy(point_k), np.copy(point_k)
+        grad_minus, grad_plus = np.copy(grad_k), np.copy(grad_k)
+        r_minus, r_plus = np.copy(r_k), np.copy(r_k)
+
+        # run NUTS
+        while (s == 1) and (j <= self.max_depth):
+            # sample a direction
+            v = int(2*(np.random.rand() < 0.5)-1)
+
+            # build tree: doubling procedure
+            if (v == -1):
+                point_minus, r_minus, grad_minus, _, _, _, \
+                    point_prime, logd_prime, grad_prime,\
+                        n_prime, s_prime, alpha, n_alpha = \
+                            self._BuildTree(point_minus, r_minus, grad_minus,
+                                            Ham, log_u, v, j, self._epsilon)
+            else:
+                _, _, _, point_plus, r_plus, grad_plus, \
+                    point_prime, logd_prime, grad_prime,\
+                        n_prime, s_prime, alpha, n_alpha = \
+                            self._BuildTree(point_plus, r_plus, grad_plus,
+                                            Ham, log_u, v, j, self._epsilon)
+
+            # Metropolis step
+            alpha2 = min(1, (n_prime/n)) #min(0, np.log(n_p) - np.log(n))
+            if (s_prime == 1) and (np.random.rand() <= alpha2):
+                self.current_point = point_prime
+                self.current_target_logd = logd_prime
+                self.current_target_grad = np.copy(grad_prime)
+                self._acc.append(1)
+            else:
+                self._acc.append(0)
+
+            # update number of particles, tree level, and stopping criterion
+            n += n_prime
+            dpoints = point_plus - point_minus
+            s = s_prime *\
+                int((dpoints @ r_minus.T) >= 0) * int((dpoints @ r_plus.T) >= 0)
+            j += 1
+            self._alpha = alpha
+            self._n_alpha = n_alpha
+
+        # update run diagnostic attributes
+        self._update_run_diagnostic_attributes(
+            self._num_tree_node, self._epsilon, self._epsilon_bar)
+        
+        self._epsilon = self._epsilon_bar 
+        if np.isnan(self.current_target_logd):
+            raise NameError('NaN potential func')
+
+    def tune(self, skip_len, update_count):
+        """ adapt epsilon during burn-in using dual averaging"""
+        k = update_count+1
+
+        # Fixed parameters that do not change during the run
+        gamma, t_0, kappa = 0.05, 10, 0.75 # kappa in (0.5, 1]
+
+        eta1 = 1/(k + t_0)
+        self._H_bar = (1-eta1)*self._H_bar +\
+            eta1*(self.opt_acc_rate - (self._alpha/self._n_alpha))
+        self._epsilon = np.exp(self._mu - (np.sqrt(k)/gamma)*self._H_bar)
+        eta = k**(-kappa)
+        self._epsilon_bar =\
+            np.exp(eta*np.log(self._epsilon) +(1-eta)*np.log(self._epsilon_bar))
+
+    def get_state(self):
+        pass
```

Comment:
Does this need updating to set and get specific keys for the state of NUTS?

Reply:
If possible I would suggest adding the keys that NUTS uses for state to the class and removing the get_state and set_state methods.

It also gives  a clearer overview of what is the actual parameters that change during sampling.

Reply:
Thank you both for pointing this out! I missed removing the `get_state` and `set_state`. Now it is removed. 

I appended keys to the STATE and HISTORY data structures. 

My plan for the state to be thoroughly tested (making sure no state variable is missing) when checkpointing is addressed #404

Reply:
Great thanks!